posts/2020/03/pytorch-tensoriterator-internals.md

```diff
@@ -0,0 +1,236 @@
+specified in the second and fourth arguments (in this case `t` and `r_`), and the dimension that
+we want to iterate over is specified as the fifth argument (`dimension`).
+
+We then follow these arguments with the main body of the iterator (which is accepted as the sixth
+argument into the macro), and denote the data, stride and size of the particular tensor dimension
+by using variables that are suffixed by `_data`, `_stride` and `_size` respectively after the
+variable name that represents the tensor inside the iterator body. For example, the size of the
+input tensor is denoted as `t_size` in the above example and the pointer to the data of the output
+tensor is denoted as `r__data`. The `accreal` in the second line is custom type that specifies
+a real number that is an accumulator (in this case for accumulating the product).
+
+Internally, the `TH_TENSOR_DIM_APPLY2` macro is expanded for generating various dispatch calls 
+depending on the type of the tensor that needs to be iterated over. The implementation of 
+`TH_TENSOR_DIM_APPLY2` can be found [here](https://github.com/pytorch/pytorch/blob/master/aten/src/TH/THTensorDimApply.h#L138).
+
+## Limitations of TH iterators
+
+Apart from the obvious complication that arises due to maintaining a codebase that is so dependent
+on such insanely complex macro expansions, TH iterators have some fundamental shortcomings. For
+one thing, they cannot be used for writing iterators in a device independent manner - you will
+need separate iterators for CPU and CUDA. Also, parallelization does not happen implcitly
+inside the iterator, you need to write the parallel looping logic yourself. Moreover, at a deeper
+level `TH` iterators do not collapse the dimensions of the tensor (as we'll see later in this
+post) therefore leading to looping that might not be as cache-optimized as possible.
+
+These limitations led to the creation of `TensorIterator`, which is used by the
+`ATen` tensor implementation for overcoming some of the shortcomings of the previous `TH`
+iterators.
+
+# Basics of TensorIterator
+
+A `TensorIterator` can be created using the default constructor. You must then add the tensors
+that you want as inputs or outputs. A good example can be found from the `TensorIterator::binary_op()`
+[method](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorIterator.cpp#L652) that
+allows you to create `TensorIterator` objects for performing point-wise binary operations
+between two tensors. The important parts look like so:
+
+``` cpp
+auto iter = TensorIterator();
+
+iter.add_output(out);
+iter.add_input(a);
+iter.add_input(b);
+
+iter.build();
+```
+As you can see, you add a tensor called `out` as the output tensors and `a` and `b` as the
+input tensors. Calling `build` is then mandatory for creating the object and letting
+the class perform other optimizations like collapsing dimesions and figuring out
+data types.
```

Comment:
Not clear to me. How does it figure out things? You mean broadcasting?

Reply:
So this was meant to be a statement for conveying that TensorIterator does so and so after calling build. The details are elaborated later. But you're right in saying that it seems a little ambiguous.